treesit-auto.el

```diff
@@ -255,65 +257,33 @@ automatic installation (or prompting, based on the value of
+  "Try to install the grammar matching the current major-mode.
 
-(defvar treesit-auto-mode-lang-alist
-  nil
-  "Lookup tree-sitter lang symbol using an Emacs major mode.")
-
-(defvar treesit-auto--original-language-source-alist
-  (purecopy treesit-language-source-alist)
-  "Keep track of `treesit-language-source-alist'.")
-
-(defvar treesit-auto--original-major-mode-remap-alist
-  (purecopy major-mode-remap-alist)
-  "Keep track of `major-mode-remap-alist'.")
-
-;; This function is the juicy, "black magic" part of the package.
-(defun treesit-auto--build-alists ()
-  "Rebuild internal alists from language recipes.
-
-Applies all of the `treesit-auto' settings to adjust
-`treesit-language-source-alist' and `major-mode-remap-alist'.
-Called whenever enabling `global-treesit-auto-mode'."
-  (setq treesit-auto-lang-recipe-alist ())
-  (setq treesit-auto-mode-lang-alist ())
-  (dolist (recipe treesit-auto-recipe-list)
-    (when-let* ((lang (treesit-auto-recipe-lang recipe))
-                (ts-mode (treesit-auto-recipe-ts-mode recipe))
-                (remap (ensure-list (treesit-auto-recipe-remap recipe)))
-                (fallback (car (seq-filter 'fboundp remap))))
-      ;; For lang -> Emacs metadata lookups
-      (push `(,lang . ,recipe) treesit-auto-lang-recipe-alist)
-      ;; For mode -> lang lookup
-      (dolist (mode remap)
-        (push `(,mode . ,lang) treesit-auto-mode-lang-alist))
-      (push `(,ts-mode . ,lang) treesit-auto-mode-lang-alist)
-      ;; Tree-sitter <--> fallback automation happens here
-      (if (treesit-auto--ready-p ts-mode)
-          (dolist (mode remap)
-            (add-to-list 'major-mode-remap-alist `(,mode . ,ts-mode)))
-        (when (fboundp fallback)
-          (add-to-list 'major-mode-remap-alist `(,ts-mode . ,fallback))))
-      ;; For `treesit-install-langauge-grammar'
-      (add-to-list 'treesit-language-source-alist
-                   `(,(treesit-auto-recipe-lang recipe)
-                     . (,(treesit-auto-recipe-url recipe)
-                        ,(treesit-auto-recipe-revision recipe)
-                        ,(treesit-auto-recipe-source-dir recipe)
-                        ,(treesit-auto-recipe-cc recipe)
-                        ,(treesit-auto-recipe-c++ recipe)))))))
+If the tree-sitter grammar is missing for the current major mode,
+this will silently fail, automatically install the grammar, or
+prompt the user about automatic installation, depending on the
+value of `treesit-auto-install'.  If installation of the grammar
+is successful, activate the tree-sitter major mode."
+  (when-let* ((not-ready (not (treesit-auto--ready-p major-mode)))
+              (recipe (treesit-auto--get-mode-recipe))
+              (ts-mode (treesit-auto-recipe-ts-mode recipe))
+              (ts-mode-exists (fboundp ts-mode))
+              (lang (treesit-auto-recipe-lang recipe)))
+    ;; install required grammars
+    (dolist (lang (or (treesit-auto-recipe-requires recipe) '()))
```

Comment:
Suggestion: what do you think of using `ensure-list` here, so that we can specify a single symbol *or* a list of symbols to require?   Much like how `:remap` can be either one mode or a list of modes https://github.com/renzmann/treesit-auto/blob/45f5d7d505745a32d8c628af49db202d56f168ba/treesit-auto.el#L287

Then we could just do `:requires 'heex` in the recipe above.

Reply:
yeah, think that might be a good change